include/nbl/builtin/hlsl/workgroup/shared_scan.hlsl

```diff
@@ -75,72 +73,59 @@ struct WorkgroupScanHead
             if(participate)
             {
                 if(any(bool2(gl_LocalInvocationIndex == lastInvocationInLevel, isLastSubgroupInvocation))) {
-                    sharedAccessor.main.set(nextLevelStoreIndex, scan);
+                    sharedAccessor.main.set(scanLoadIndex - loadStoreIndexDiff, scan);
                 }
             }
             sharedAccessor.main.workgroupExecutionAndMemoryBarrier();
             participate = gl_LocalInvocationIndex <= (lastInvocationInLevel >>= glsl::gl_SubgroupSizeLog2());
             if(participate)
             {
-                const uint prevLevelScan = sharedAccessor.main.get(gl_LocalInvocationIndex);
+                const uint prevLevelScan = sharedAccessor.main.get(scanLoadIndex);
                 scan = subgroupOp(prevLevelScan);
-                if(isScan) {
-                    sharedAccessor.main.set(scanStoreIndex, scan);
-                }
             }
         }
-        return scan;
+        lastLevelScan = scan; // only invocations of gl_LocalInvocationIndex < gl_SubgroupSize will have correct values, rest will have garbage
     }
 };
 
-template<typename T, class Binop, class SharedAccessor>
+template<typename T, class Binop, class ScanHead, class SharedAccessor, bool isExclusive>
 struct WorkgroupScanTail
 {
-    bool isExclusive;
     T identity;
-    T firstLevelScan;
-    uint lastInvocation;
-    uint scanStoreIndex;
+    ScanHead head;
 
-    static WorkgroupScanTail create(bool isExclusive, T identity, T firstLevelScan, uint lastInvocation, uint scanStoreIndex)
+    static WorkgroupScanTail create(T identity, ScanHead head)
     {
-        WorkgroupScanTail<T, Binop, SharedAccessor> wst;
-        wst.isExclusive = isExclusive;
-        wst.identity = identity;
-        wst.firstLevelScan = firstLevelScan;
-        wst.lastInvocation = lastInvocation;
-        wst.scanStoreIndex = scanStoreIndex;
-        return wst;
+        WorkgroupScanTail<T, Binop, ScanHead, SharedAccessor, isExclusive> tail;
+        tail.identity = identity;
+        tail.head = head;
+        return tail;
     }
 
-    T operator()()
+    T operator()(NBL_REF_ARG(SharedAccessor) sharedAccessor)
     {
         Binop binop;
-        SharedAccessor sharedAccessor;
-        sharedAccessor.main.workgroupExecutionAndMemoryBarrier();
-        
+        uint lastInvocation = head.lastInvocation;
+        uint firstLevelScan = head.firstLevelScan;
         const uint subgroupId = glsl::gl_SubgroupID();
+        
         if(lastInvocation >= glsl::gl_SubgroupSize())
         {
-            uint scanLoadIndex = scanStoreIndex + glsl::gl_SubgroupSize();
+            uint scanLoadIndex = head.scanLoadIndex + glsl::gl_SubgroupSize();
             const uint shiftedInvocationIndex = gl_LocalInvocationIndex + glsl::gl_SubgroupSize();
             const uint currentToHighLevel = subgroupId - shiftedInvocationIndex;
+            
+            sharedAccessor.main.set(head.scanLoadIndex, head.lastLevelScan);
```

Comment:
this is gonna become an OOB `set`, you need to wrap it in an `if (participate)` and have the `participate be a left over from the Head

Reply:
you also don't want to do any scrolling/adjustments to `scanLoadIndex` you just want a barrier before the `if (participate)` because you can store the scanned elements right back to the same memory location you read from

Reply:
ok here's a legibility update (replace the whole `if(lastInvocation >= glsl::gl_SubgroupSize())` block)
```cpp
    // this is one less than in the Upsweep cause we need to translate high level inclusive scans into exclusive on the fly
    // btw due to 2s complement the overflow present in the first invocation doesn't matter
    const uint32_t storeLoadIndexDiff = gl_LocalInvocationIndex - glsl::gl_SubgroupID() - 1;
    
    // because DXC doesn't do references and I need my "frozen" registers
    #define scanStoreIndex head.scanLoadIndex
    // as we iterate some previously active (higher level) invocations hold their exclusive prefix sum in `lastLevelScan`
    while (not lowest level)
    {
        if (participate)
        {
            // we need to add the higher level invocation exclusive prefix sum to current value
            if (not first iteration) // but the top level doesn't have any level above itself
                 head.lastLevelScan = binop(head.lastLevelScan,sharedAccessor.get(scanStoreIndex));
            // now `lastLevelScan` has current level's inclusive prefux sum computed properly
            // note we're overwriting the same location with same invocation so no barrier needed
            // we store everything even though we'll never use the last entry due to shuffleup on read
            sharedAccessor.set(scanStoreIndex,head.lastLevelScan);
        }
        sharedAccessor.main.workgroupExecutionAndMemoryBarrier();
        participate = ???????????????????????; // widen
        if (participate)
        {
            // we either need to prevent OOB read altogether OR cmov identity after the far
            head.lastLevelScan = bool(gl_LocalInvocationIndex) ? Binop::identity:sharedAccessor.get(scanStoreIndex-storeLoadIndexDiff);
            // this is fine if on the way up you also += under `if (participate)`
            scanStoreIndex -= lastInvocationInLevel+1;
        }
    }
    #undef scanStoreIndex
    
    // the very first prefix sum we did is in a register, not Accessor scratch mem hence the special path
    if(lastInvocation >= glsl::gl_SubgroupSize())
         firstLevelScan = binop(head.lastLevelScan,firstLevelScan);
```